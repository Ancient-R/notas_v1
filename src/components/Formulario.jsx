import React, { useState } from 'react'
import uuid from 'uuid/dist/v4'
import Error from './Error'

const Formulario =({ nota, setNota, notas, setNotas, nuevocontenido }) => {

    const { titulo, contenido, importancia } = nota
    const [ error, setError ] = useState(false)

    const actualizarState = e =>{
        setNota({
            ...nota, [e.target.name] : e.target.value
        })
    }

    const agregarNota = e =>{
        e.preventDefault()

        const regular= /[a-zA-Z0-9]/

        if(!regular.test(titulo) || contenido === ''){ 
            setError(true)
            return
        }

        setError(false)

        //Se agrega un id a la nota
        nota.id = uuid()

        //Se crea la nota y se añade al array de citas
        setNotas([
            ...notas, nota
        ])

        //Se reinicia el formulario
        setNota({
            titulo: '',
            contenido: '',
            importancia: 'low'
        })
        
    }

    return ( 
        <form
            className="mt-5"
            onSubmit = { agregarNota }
        >
            {error ?  <Error 
                        error="Debe llenar todos los campos"/>
                    : null
            }

            <div className="input-group">
                <div className="input-group-prepend">
                    <span className="input-group-text bg-primary text-white titulo">Título <i className="fas fa-text-height ml-2"/> </span>
            </div>
            <input 
                type="text" 
                className="form-control"
                name="titulo"
                value = { titulo }
                onChange={ actualizarState }/>
            </div>
            <small className="text-muted">Ejemplo: Tarea</small>
            <div className="input-group mt-3">
                <div className="input-group-prepend">
                    <span className="input-group-text bg-primary text-white">Contenido <i className="fas fa-align-center ml-2" /> </span>
            </div>
            <textarea 
                name="contenido"
                rows="2"
                className="form-control text-area"
                value={contenido}
                onChange={ actualizarState }/>
            </div>
            <small className="text-muted">Ejemplo: Investigar cómo nacen los bebés</small>

            <div className="input-group mt-3">
                <div className="input-group-prepend">
                    <span className="input-group-text bg-primary text-white">Importancia
                    <i className="fas fa-exclamation ml-2"></i></span>
                </div>
                <select 
                    name="importancia" className="form-control"
                    value={importancia}
                    onChange={ actualizarState }
                >
                    <option value="low" className="bg-secondary text-white">Baja</option>
                    <option value="media" className="alert-warning">Media</option>
                    <option value="higth" className="bg-danger text-white">Alta</option>
                </select>
            </div>
            <small className="text-muted mb-3">Por defecto la importancia será 'baja'</small>
            <div className="input-group mt-3 mb-3">
                <button 
                    type="submit"
                    value="Agregar nota"
                    className="btn btn-primary btn-block"
                >Agregar nota <span><i className="far fa-sticky-note ml-2"/></span></button>
            </div>
        </form>
    );
}

export default Formulario;