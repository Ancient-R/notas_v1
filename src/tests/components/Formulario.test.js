import React from 'react'
import { shallow } from 'enzyme';

// component
import Formulario from '../../components/Formulario';

import '@testing-library/jest-dom';

const nota = {
    id: 1234,
    titulo: '',
    contenido: '',
    importancia: 'low'
}

const setNota = jest.fn();
const notas = [];
const setNotas = jest.fn();
const nuevocontenido = '';

const wrapper = shallow(
    <Formulario
        nota={ nota }
        notas={ notas }
        setNota={ setNota }
        setNotas={ setNotas }
        nuevocontenido={ nuevocontenido }
     />
);

describe('Pruebas en el componente <Formulario />', () => {
    test('debe mostrarse correctamente', () => {

        expect( wrapper ).toMatchSnapshot();
    });
    test('debe ejecutar setNota', () => {

        wrapper.find('input[name="titulo"]').simulate('change', {
            target: {
                name: 'titulo',
                value: 'Título test'
            }
        });

        expect( setNota ).toHaveBeenCalledWith({
            id: 1234, 
            titulo: "Título test",
            contenido: "",
            importancia: 
            "low",
        });
    });

    test('no debe ejecutar setNotas', () => {
        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });
        expect( setNotas ).not.toHaveBeenCalled();
    });

    test('debe ejecutar setNotas', () => {
       
        const nota = {
            id: 1234,
            titulo: 'Título testing',
            contenido: 'Contenido del título testing',
            importancia: 'low'
        }

        const setNota = jest.fn();
        const notas = [];
        const setNotas = jest.fn();
        const nuevocontenido = '';

        const wrapper = shallow(
            <Formulario
                nota={ nota }
                notas={ notas }
                setNota={ setNota }
                setNotas={ setNotas }
                nuevocontenido={ nuevocontenido }
            />
        );


        wrapper.find('input[name="titulo"]').simulate('change', {
            target: {
                name: 'titulo',
                value: 'Título testing'
            }
        });

        wrapper.find('textarea').simulate('change', {
            target: {
                name: 'contenido',
                value: 'Contenido del título testing'
            }
        });

        wrapper.find('select').simulate('change', {
            target:{
                name: 'importancia',
                value: 'low'
            }
        })

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        expect( setNotas ).toHaveBeenCalledWith([{
            id: expect.any(String),
            titulo: "Título testing",
            contenido: "Contenido del título testing", 
            importancia: "low"
        }]);
        });
});