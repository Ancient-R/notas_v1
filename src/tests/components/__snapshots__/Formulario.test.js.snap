// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <Formulario /> debe mostrarse correctamente 1`] = `
<form
  className="mt-5"
  onSubmit={[Function]}
>
  <div
    className="input-group"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text bg-primary text-white titulo"
      >
        Título 
        <i
          className="fas fa-text-height ml-2"
        />
         
      </span>
    </div>
    <input
      className="form-control"
      name="titulo"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <small
    className="text-muted"
  >
    Ejemplo: Tarea
  </small>
  <div
    className="input-group mt-3"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text bg-primary text-white"
      >
        Contenido 
        <i
          className="fas fa-align-center ml-2"
        />
         
      </span>
    </div>
    <textarea
      className="form-control text-area"
      name="contenido"
      onChange={[Function]}
      rows="2"
      value=""
    />
  </div>
  <small
    className="text-muted"
  >
    Ejemplo: Investigar cómo nacen los bebés
  </small>
  <div
    className="input-group mt-3"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text bg-primary text-white"
      >
        Importancia
        <i
          className="fas fa-exclamation ml-2"
        />
      </span>
    </div>
    <select
      className="form-control"
      name="importancia"
      onChange={[Function]}
      value="low"
    >
      <option
        className="bg-secondary text-white"
        value="low"
      >
        Baja
      </option>
      <option
        className="alert-warning"
        value="media"
      >
        Media
      </option>
      <option
        className="bg-danger text-white"
        value="higth"
      >
        Alta
      </option>
    </select>
  </div>
  <small
    className="text-muted mb-3"
  >
    Por defecto la importancia será 'baja'
  </small>
  <div
    className="input-group mt-3 mb-3"
  >
    <button
      className="btn btn-primary btn-block"
      type="submit"
      value="Agregar nota"
    >
      Agregar nota 
      <span>
        <i
          className="far fa-sticky-note ml-2"
        />
      </span>
    </button>
  </div>
</form>
`;
